analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Invocation.cs

```diff
@@ -1082,10 +1082,110 @@ public void Main()
 private static bool Equals() => false;
 private static bool Equals(object a, object b, object c) => false;";
             var validator = SETestContext.CreateCSMethod(code).Validator;
-            validator.ValidateTag("InstanceOne", x => x.AllConstraints.Should().ContainSingle().Which.Should().Be(ObjectConstraint.NotNull));
-            validator.ValidateTag("InstanceTwo", x => x.AllConstraints.Should().ContainSingle().Which.Should().Be(ObjectConstraint.NotNull));
-            validator.ValidateTag("NoArgs", x => x.AllConstraints.Should().ContainSingle().Which.Should().Be(ObjectConstraint.NotNull));
-            validator.ValidateTag("MoreArgs", x => x.AllConstraints.Should().ContainSingle().Which.Should().Be(ObjectConstraint.NotNull));
+            validator.ValidateTag("InstanceOne", x => x.Should().HaveOnlyConstraint(ObjectConstraint.NotNull));
+            validator.ValidateTag("InstanceTwo", x => x.Should().HaveOnlyConstraint(ObjectConstraint.NotNull));
+            validator.ValidateTag("NoArgs", x => x.Should().HaveOnlyConstraint(ObjectConstraint.NotNull));
+            validator.ValidateTag("MoreArgs", x => x.Should().HaveOnlyConstraint(ObjectConstraint.NotNull));
+        }
+
+        [DataTestMethod]
+        [DataRow("null", "null", true, ConstraintKind.ObjectNull, ConstraintKind.ObjectNull)]
+        [DataRow("null", "new object()", false, ConstraintKind.ObjectNull, ConstraintKind.ObjectNotNull)]
+        [DataRow("new object()", "null", false, ConstraintKind.ObjectNotNull, ConstraintKind.ObjectNull)]
+        [DataRow("new int?()", "null", true, ConstraintKind.ObjectNull, ConstraintKind.ObjectNull)]
+        [DataRow("(int?)null", "null", true, ConstraintKind.ObjectNull, ConstraintKind.ObjectNull)]
+        [DataRow("new int?(42)", "null", false, ConstraintKind.ObjectNotNull, ConstraintKind.ObjectNull)]
+        [DataRow("new int?()", "42", false, ConstraintKind.ObjectNull, ConstraintKind.ObjectNotNull)]
+        public void Invocation_ReferenceEquals_LearnResult(string left, string right, bool expectedResult, ConstraintKind expectedConstraintLeft, ConstraintKind expectedConstraintRight)
+        {
+            var code = $"""
+                object left = {left};
+                object right = {right};
+                var result = ReferenceEquals(left, right);
+                Tag("Result", result);
+                Tag("Left", left);
+                Tag("Right", right);
+                """;
+            var validator = SETestContext.CreateCS(code).Validator;
+            validator.ValidateTag("Result", x => x.HasConstraint(BoolConstraint.From(expectedResult)).Should().BeTrue());
+            validator.ValidateTag("Left", x => x.AllConstraints.Select(x => x.Kind).Should().ContainSingle().Which.Should().Be(expectedConstraintLeft));
```

Comment:
Yes, it would. More precisely just another overload on `HaveOnlyConstraint`.

We just didn't have time to do it yet, the current extensions were merged this sprint.